csharp/src/Ice/MetricsObserver.cs

```diff
@@ -0,0 +1,230 @@
+//
+// Copyright (c) ZeroC, Inc. All rights reserved.
+//
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using ZeroC.Ice;
+
+namespace ZeroC.IceMX
+{
+    public partial class Metrics
+    {
+        internal object Clone() => MemberwiseClone();
+    }
+    internal class MetricsHelper<T> where T : Metrics
+    {
+        private readonly AttributeResolver _attributes;
+
+        public virtual void InitMetrics(T metrics)
+        {
+            // Override in specialized helpers.
+        }
+
+        internal string Resolve(string attribute) => _attributes.Resolve(this, attribute);
+        protected MetricsHelper(AttributeResolver attributes) => _attributes = attributes;
+
+        protected virtual string? DefaultResolve(string attribute) => null;
+
+        internal class AttributeResolver
```

Comment:
The class declaration should be moved at the top?

Reply:
fixed